firedrake/cython/dmcommon.pyx

```diff
@@ -1763,80 +1761,48 @@ def mark_entity_classes_using_cell_dm(PETSc.DM swarm):
-    labels[core_idxs] = 1
-    labels[owned_idxs] = 2
-    labels[ghost_idxs] = 3
-
-    # We know the parent DM cell index for each of our swarm points. We can
-    # therefore filter the list of all labels to find the corresponding label
-    # of each swarm point.
-    swarm_plex_cells = swarm.getField("DMSwarm_cellid")
-    swarm.restoreField("DMSwarm_cellid")
-    swarm_parent_cell_labels = labels[swarm_plex_cells]
-    assert len(swarm_parent_cell_labels) == len(swarm_plex_cells)
-    for label_idx, label in enumerate(swarm_parent_cell_labels):
-        # We set the label using label index since this index is shared across
-        # all DMSwarm fields: label index n into a given field (such as
-        # DMSwarmPIC_coor) always corresponds to the same point in the swarm.
-        if label == 1:
-            CHKERR(DMLabelSetValue(swarm_label_core, label_idx, 1))
-        elif label == 2:
-            CHKERR(DMLabelSetValue(swarm_label_owned, label_idx, 1))
-        elif label == 3:
-            CHKERR(DMLabelSetValue(swarm_label_ghost, label_idx, 1))
-        else:
-            raise RuntimeError("Unknown label value")
-    return
+    get_height_stratum(plex.dm, 0, &cStart, &cEnd)
+    CHKERR(PetscMalloc1(cEnd - cStart, &plex_cell_classes))
+    for c in range(cStart, cEnd):
+        plex_cell_classes[c - cStart] = -1
+    for ilabel, op2class in enumerate([b"pyop2_core", b"pyop2_owned", b"pyop2_ghost"]):
+        CHKERR(DMGetLabel(plex.dm, op2class, &plex_label))
+        CHKERR(DMLabelGetStratumSize(plex_label, label_value, &class_size))
+        if class_size > 0:
+            CHKERR(DMLabelGetStratumIS(plex_label, label_value, &class_is))
+            CHKERR(ISGetIndices(class_is, &class_indices))
+            for i in range(class_size):
+                if cStart <= class_indices[i] < cEnd:
+                    plex_cell_classes[class_indices[i] - cStart] = ilabel
+            CHKERR(ISRestoreIndices(class_is, &class_indices))
+            CHKERR(ISDestroy(&class_is))
+    for c in range(cStart, cEnd):
+        if plex_cell_classes[c - cStart] < 0:
+            raise RuntimeError("Cell point %d in the parent plex does not belong to any pyop2 class" % c)
+    for ilabel, op2class in enumerate([b"pyop2_core", b"pyop2_owned", b"pyop2_ghost"]):
+        CHKERR(DMCreateLabel(swarm.dm, op2class))
+        CHKERR(DMGetLabel(swarm.dm, op2class, &swarm_labels[ilabel]))
+    CHKERR(DMSwarmGetField(swarm.dm, b"DMSwarm_cellid", &blocksize, &ctype, <void**>&swarmParentCells))
+    assert ctype == PETSC_INT
+    assert blocksize == 1
+    CHKERR(DMSwarmGetLocalSize(swarm.dm, &nswarmCells))
+    for swarmCell in range(nswarmCells):
+        plex_cell_class = plex_cell_classes[swarmParentCells[swarmCell] - cStart]
+        CHKERR(DMLabelSetValue(swarm_labels[plex_cell_class], swarmCell, label_value))
+    CHKERR(DMSwarmRestoreField(swarm.dm, b"DMSwarm_cellid", &blocksize, &ctype, <void**>&swarmParentCells))
+    CHKERR(PetscFree(plex_cell_classes))
```

Comment:
So why did you rewrite this? What advantages does this rewrite have over the previous routine?

Reply:
It is more concise. It also avoid copying arrays (`.getIndices()`). This function is only called once, so efficiency does not really matter, but this is more consistent with other Cython functions in dmcommon.

Reply:
If you are interested, here's the performance difference (`temp.py`):
```
from firedrake import *
import numpy as np
from firedrake.cython import dmcommon
import time


mesh = UnitSquareMesh(200, 200)
vertexcoords = np.random.random((2000, 2))
vm = VertexOnlyMesh(mesh, vertexcoords, missing_points_behaviour=None)
swarm = vm.topology_dm

vm.comm.barrier()

start = time.time()
dmcommon.mark_entity_classes_using_cell_dm_old(swarm)
end = time.time()
print(f"original (rank = {vm.comm.rank}): {end - start} [s]", flush=True)

vm.comm.barrier()

start = time.time()
dmcommon.mark_entity_classes_using_cell_dm(swarm)
end = time.time()
print(f"updated  (rank = {vm.comm.rank}): {end - start} [s]", flush=True)
```
```
mpiexec -n 4 python temp.py
```
```
original (rank = 0): 0.000461578369140625 [s]
original (rank = 1): 0.0004367828369140625 [s]
original (rank = 2): 0.00045990943908691406 [s]
original (rank = 3): 0.0004508495330810547 [s]
updated  (rank = 0): 0.00015306472778320312 [s]
updated  (rank = 1): 0.0001506805419921875 [s]
updated  (rank = 2): 0.00015020370483398438 [s]
updated  (rank = 3): 0.00015211105346679688 [s]
```
`mark_entity_classes_using_cell_dm_old` is the original implementation.